File: frontend/lib/sync.ts
/**
 * API client wrapper for the automatic transaction sync feature
 * (backend: app.api.v1.sync). This connects to a MockSyncProvider only -
 * see app.sync.provider - never a real bank, Account Aggregator, or other
 * external financial provider. Every field here mirrors
 * backend/app/schemas/sync.py exactly; nothing here ever carries a PIN,
 * CVV, OTP, password, or provider credential of any kind - the backend
 * itself never returns one (see app.services.sync_service).
 */
import { apiRequest } from "@/lib/api-client";

export const SYNC_CONSENT_STATUSES = [
  "pending",
  "active",
  "paused",
  "revoked",
  "expired",
] as const;
export type SyncConsentStatus = (typeof SYNC_CONSENT_STATUSES)[number];

export const SYNC_RUN_STATUSES = ["success", "partial", "failed"] as const;
export type SyncRunStatus = (typeof SYNC_RUN_STATUSES)[number];

export interface LinkedAccount {
  id: string;
  account_id: string | null;
  provider: string;
  external_institution_name: string;
  fip_reference: string | null;
  consent_status: SyncConsentStatus;
  consent_expires_at: string | null;
  masked_account_ref: string | null;
  last_synced_at: string | null;
  last_sync_status: string | null;
  created_at: string;
  updated_at: string;
}

/** `consent_handle` is a short-lived, opaque token - never a credential -
 * that must be passed straight back to completeLink to finish the mock
 * connection flow. Callers must never store this beyond that. */
export interface LinkInitiation {
  provider: string;
  consent_handle: string;
  redirect_url: string;
  expires_at: string;
}

export interface SyncRun {
  id: string;
  linked_account_id: string;
  started_at: string;
  completed_at: string | null;
  status: SyncRunStatus;
  transactions_fetched: number;
  transactions_created: number;
  transactions_skipped_duplicate: number;
  error_message: string | null;
}

export async function initiateLink(accessToken: string): Promise<LinkInitiation> {
  return apiRequest<LinkInitiation>("/api/v1/sync/links", {
    method: "POST",
    accessToken,
    body: {},
  });
}

export async function completeLink(
  accessToken: string,
  consentHandle: string
): Promise<LinkedAccount[]> {
  return apiRequest<LinkedAccount[]>(
    `/api/v1/sync/links/${encodeURIComponent(consentHandle)}/callback`,
    { method: "POST", accessToken }
  );
}

export async function listLinkedAccounts(accessToken: string): Promise<LinkedAccount[]> {
  return apiRequest<LinkedAccount[]>("/api/v1/sync/links", { accessToken });
}

export async function revokeLink(
  accessToken: string,
  linkedAccountId: string
): Promise<LinkedAccount> {
  return apiRequest<LinkedAccount>(`/api/v1/sync/links/${linkedAccountId}`, {
    method: "DELETE",
    accessToken,
  });
}

export async function triggerSync(
  accessToken: string,
  linkedAccountId: string
): Promise<SyncRun> {
  return apiRequest<SyncRun>(`/api/v1/sync/links/${linkedAccountId}/sync`, {
    method: "POST",
    accessToken,
  });
}

export async function listSyncRuns(
  accessToken: string,
  linkedAccountId: string
): Promise<SyncRun[]> {
  return apiRequest<SyncRun[]>(`/api/v1/sync/links/${linkedAccountId}/runs`, { accessToken });
}

File: backend/app/schemas/sync.py
"""Linked-account/sync-run request/response schemas, used by
app.api.v1.sync.

Every request schema is `extra="forbid"`, same convention as every other
schema in this app - and deliberately carries NOTHING that could resemble a
payment credential: no PIN, CVV, OTP, password, or payment-authorization
field of any kind, ever. `LinkedAccountUpdate` in particular only ever
accepts the one field app.services.sync_service actually supports mutating
(`account_id`) - a client can never touch consent_id, provider identifiers,
sync statistics, ownership, or consent_status through it.

Every response schema is equally deliberate about what it leaves OUT: no
provider consent secret, no full raw provider payload, no masked-account
value beyond what app.models.linked_account itself ever stores (already
masked at the source).
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.linked_account import SyncConsentStatus
from app.models.sync_run import SyncRunStatus


class LinkInitiateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    # A free-text hint only (e.g. "HDFC Bank") - never a credential, and
    # the mock provider does not even use it today; see
    # app.sync.provider.base.BankSyncProvider.initiate_link.
    institution_hint: str | None = Field(default=None, max_length=200)


class LinkInitiationRead(BaseModel):
    """`consent_handle` is an opaque, short-lived token the client must
    pass back to the callback endpoint - never a credential or a secret
    the provider could use to move money; see
    app.sync.provider.base.LinkInitiation."""

    model_config = ConfigDict(from_attributes=True)

    provider: str
    consent_handle: str
    redirect_url: str
    expires_at: datetime


class LinkedAccountUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    account_id: UUID


class LinkedAccountRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    account_id: UUID | None
    provider: str
    external_institution_name: str
    fip_reference: str | None
    consent_status: SyncConsentStatus
    consent_expires_at: datetime | None
    masked_account_ref: str | None
    last_synced_at: datetime | None
    last_sync_status: str | None
    created_at: datetime
    updated_at: datetime


class SyncRunRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    linked_account_id: UUID
    started_at: datetime
    completed_at: datetime | None
    status: SyncRunStatus
    transactions_fetched: int
    transactions_created: int
    transactions_skipped_duplicate: int
    error_message: str | None
